File: 0049_group_anagrams/frequency_map/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <array>
#include <cassert>

// Custom hash for std::array<int, 26> to use it as an unordered_map key
struct ArrayHasher {
    std::size_t operator()(const std::array<int, 26>& a) const {
        std::size_t h = 0;
        for (int x : a) {
            h ^= std::hash<int>{}(x) + 0x9e3779b9 + (h << 6) + (h >> 2);
        }
        return h;
    }
};

class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) {
        std::unordered_map<std::array<int, 26>, std::vector<std::string>, ArrayHasher> groups;
        
        for (const std::string& s : strs) {
            std::array<int, 26> count = {0};
            for (char c : s) {
                count[c - 'a']++;
            }
            groups[count].push_back(s);
        }
        
        std::vector<std::vector<std::string>> result;
        for (auto& pair : groups) {
            result.push_back(std::move(pair.second));
        }
        return result;
    }
};

void test() {
    Solution sol;
    std::vector<std::string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    auto result = sol.groupAnagrams(strs);
    assert(result.size() == 3);
    std::cout << "All tests passed!" << std::endl;
}

int main() {
    test();
    return 0;
}

File: 0049_group_anagrams/frequency_map/go/solution.go
package frequency_map

func groupAnagrams(strs []string) [][]string {
	// In Go, arrays are comparable and can be used as map keys
	groups := make(map[[26]int][]string)
	
	for _, s := range strs {
		var count [26]int
		for _, char := range s {
			count[char-'a']++
		}
		groups[count] = append(groups[count], s)
	}
	
	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}
	return result
}
